s2/s2_test.go

```diff
@@ -41,6 +46,13 @@ const (
 	epsilon = 1e-15
 )
 
+var (
+	// set up a random generator for use in tests.
+	// Prior to Go 1.20, the generator was seeded like Seed(1) at program startup.
+	// Using the default seed flag value will produce the prior behavior.
+	random = rand.New(rand.NewSource(*s2RandomSeed))
```

Comment:
How is this going to interact with `test_filter`? If there's a failure, we don't want it to go away when we change the test filter.

It seems we don't want one random source shared with all test cases, it should be re-initialized in each one.

Reply:
> Otherwise we would need to go into every single Test... function in every file and add the NewSource(...) line to each method to do this. For most tests that's overkill since they really don't care, they usually just need whatever random value comes next.

This is what we did in C++. More flaky tests will almost certainly be found this way. It's a bigger project, though.

Reply:
So far it seems to have highlighted a couple more while I think I've tracked down what was wrong with a couple of the others.